crux-dbg/src/Crux/Debug/LLVM.hs

```diff
@@ -150,9 +172,74 @@ do_breakpoint :: (wptr ~ CLE.ArchWidth arch)
+             -> C.OverM personality sym (CLI.LLVM arch) String
+lookupString mvar ptr =
+  do sym <- LCS.getSymInterface
+     mem <- LCS.readGlobal mvar
+     bytes <- liftIO (CLM.loadString sym mem (regValue ptr) Nothing)
+     return (BS8.unpack (BS.pack bytes))
+
+-- TODO: export in crux
+withSolverAdapter :: CCS.SolverOffline -> (WS.SolverAdapter st -> a) -> a
+withSolverAdapter solverOff k =
+  case solverOff of
+    CCS.Boolector -> k WS.boolectorAdapter
+    CCS.DReal -> k WS.drealAdapter
+    CCS.SolverOnline CCS.CVC4 -> k WS.cvc4Adapter
+    CCS.SolverOnline CCS.STP -> k WS.stpAdapter
+    CCS.SolverOnline CCS.Yices -> k WS.yicesAdapter
+    CCS.SolverOnline CCS.Z3 -> k WS.z3Adapter
+
+do_debug_assert :: ( CLO.ArchOk arch
+                  , LCB.IsSymInterface sym
+                  , CLM.HasLLVMAnn sym
+                  , sym ~ (WEB.ExprBuilder t st fs)
+                  , SC.Architecture arch' t
+                  , SC.SymbolicArchitecture arch' t
+                  , ext ~ CLI.LLVM arch
+                  )
+                => C.CruxOptions
+                -> SB.DebuggerConfig t ext arch'
+                -> LCS.GlobalVar CLM.Mem
+                -> sym
+                -> Ctx.Assignment (LCS.RegEntry sym) (Ctx.EmptyCtx Ctx.::> LCT.BVType 8 Ctx.::> CLT.LLVMPointerType (CLE.ArchWidth arch) Ctx.::> LCT.BVType 32)
+                -> C.OverM personality sym (CLI.LLVM arch) (LCS.RegValue sym LCT.UnitType)
+do_debug_assert cruxOpts sconf mvar sym (Ctx.Empty Ctx.:> p Ctx.:> pFile Ctx.:> line) = do
+  cond <- liftIO $ bvIsNonzero sym (regValue p)
+  file <- lookupString mvar pFile
+  l <- case asBV (regValue line) of
+         Just (BV.BV l)  -> return (fromInteger l)
+         Nothing -> return 0
+  let pos = WPL.SourcePos (T.pack file) l 0
+  loc <- liftIO $ getCurrentProgramLoc sym
+  let loc' = loc{ WPL.plSourceLoc = pos }
+  let msg = LCS.GenericSimError "crucible_debug_assert"
+  ret <- liftIO $ LCB.addDurableAssertion sym (LCB.LabeledPred cond (LCS.SimError loc' msg))
+  let offSolver = case CCS.parseSolverConfig cruxOpts of
+                    Right (CCS.SingleOnlineSolver _onSolver) -> CCS.SolverOnline _onSolver
+                    Right (CCS.OnlineSolverWithOfflineGoals _ _offSolver) -> _offSolver
+                    Right (CCS.OnlyOfflineSolver _offSolver) -> _offSolver
+                    Right (CCS.OnlineSolverWithSeparateOnlineGoals _ _onSolver) -> CCS.SolverOnline _onSolver
+                    Left _ -> CCS.SolverOnline CCS.Yices
+  withSolverAdapter offSolver $ \adapter -> do
+    let logData = WS.defaultLogData
+    pathCond <- liftIO $ LCB.getPathCondition sym
+    satTest <- liftIO $ solver_adapter_check_sat adapter sym logData [pathCond, cond] $ \satRes ->
+      case satRes of
+        Sat _ -> return True
+        _ -> return False
+    simState <- CMS.get
+    let ng = WEB.exprCounter sym
+    liftIO $ when (not satTest) . void $ SB.surveyorState sconf ng simState Nothing
```

Comment:
I'd have to work this out on paper, but I think the safer check here is something like:
```
negCond <- notPred cond
solver_adapter_check_sat .. [pathCond, cond] $ \satRes ->
  case satRes of
    Unsat -> do
      -- Do nothing, the assertion can never fail for any input
      return ret
   Sat model -> do
      -- start surveyor
```

I think the logic you currently have with `Sat -> True` and negating `satTest`  works when the path condition is empty, but I don't think it is correct in the presence of a non-trivial path condition.

Reply:
I refactored the check and also tested it on a few example programs with conditionals which should have a non-trivial path condition.